### ISSUE TITLE
Create linked-tabs widget for displaying platform-specific instructions

### ISSUE THREAD

**Issue by dedemorton** [elastic/docs#260]
We need a better way of displaying platform-specific instructions (especially commands) in the docs.

It would be nice to have a widget with tabs or some other mechanism that allows users to select a platform and see the correct commands/instructions for that platform. 

---
elastic/beats#17007 added lots of custom pass-through HTML to support linked tab widgets.

Add an Asciidoctor extension/directive to make it easier to create these on a page.

Depends on #1895

**Comment by sophiec20** [comment]
Just seen this... might not be applicable to us, but still pleasing to the eye imho.
https://support.google.com/a/answer/7491144?hl=en

**Comment by dedemorton** [comment]
@nik9000 I'm wondering if we can get some traction on this. For Beats, we provide users with platform-specific commands for getting started, but as we add more ways to install/manage beats, our instructions are getting very long...for very simple tasks.

I'd love to see something slick, but even a crude tabbed widget would be better than what we have now. I mocked one up in HTML in about 20 min. Not proposing that we use this, but it is possible to get something just by wrapping content in HTML (in this case, radio buttons!) and adding some CSS styles: https://oswidget-9b394.firebaseapp.com/packetbeat-starting.html

My example is not meant to be good design, and it won't work on older browsers, but it does show that we can do something like this with HTML and CSS changes.

I would rather have something really slick, but we've been waiting for a redesign of the website for years now.

**Comment by nik9000** [comment]
I'm still on vacation for a next two weeks but some kind of platform
selecting widget does sound like a good idea to me.

**Comment by dedemorton** [comment]
@gtback This is one of the enhancements that we need for solution-based docs.

@bmorelli25 Created a basic prototype that demonstrates the kind of mechanism we need: 
https://github.com/elastic/apm-server/issues/3556

Brandon recently pointed me towards an asciidoctor extension that adds tabs, but I don't know how difficult it would be to use this in our current build environment.

* https://stackoverflow.com/questions/38211766/using-tabs-in-asciidoc-spring-rest-docs
* https://github.com/spring-io/spring-asciidoctor-extensions/blob/master/README.adoc

Can you take a look and see if it's feasible to add this extension to our build somehow?

I've been waiting for a long time for this enhancement. I hope we can get some traction soon

**Comment by gtback** [comment]
@dedemorton Part of the reason for the complexity of the current approach for code snippets is that content is sourced from other repos (generally, Elasticsearch client libraries). Would that be the case here, or would it be (like in @bmorelli25's example) all drawn from the current repo and just using tabs as a visual/UX improvement?

I don't know we could use that extension directly, because it seems to be targeting the Java implementation of Asciidoctor, but I'm sure we could find (or write) something similar ourselves, or even adapt and reuse the existing widget if it's OK.

**Comment by dedemorton** [comment]
>Would that be the case here, or would it be (like in @bmorelli25's example) all drawn from the current repo and just using tabs as a visual/UX improvement?

For Beats, everything lives in one repo right now. That might change as we roll out some of the content reuse we've discussed.

So ... did a little experimenting yesterday and came up with a prototype for Beats based on the work that @bmorelli25 did for APM:  My changes are mixed in with other content changes that are in progress. And I offer the disclaimer that I'm not a web developer, but here it is:

https://github.com/elastic/beats/pull/17007 (widget code is under docs/tab-widgets).

I used Brandon's code and slapped a different CSS on it. The CSS was lifted from W3C schools, but I figured we would want to replace it with our own styles anyhow.

Again, I kind of hacked this together. Ideally we would want the build to generate the widget code based on information that's specified in the tagged regions.

I wonder if we can use this hacky way forward for now and plan to replace it with something that's more automated. The main requirement, I guess, would be that we find a way to use the tagged regions. That would help us avoid any rework.

**Comment by dedemorton** [comment]
Update on the tabbed panels in the Beats documentation:

For now, we are maintaining the tabbed panel code manually and injecting it into the HTML files. I just merged a PR that introduces the widgets here: https://github.com/elastic/beats/pull/17007. They are going into the 7.9 release. 

It would be great at some point to be able to generate these widgets.

**Comment by dedemorton** [comment]
We are considering our "widget" code (seen [here](https://www.elastic.co/guide/en/beats/filebeat/master/filebeat-installation-configuration.html#installation)) to be experimental until we figure out the next steps. We need to get some of this figured out before propagating these widgets to more pages:

-  Get design/developer resources to improve the javascript and css code. Right now, the tabbed panels display fine on desktop browsers, but aren't optimized for mobile (though that's true of our docs, too).
-  Add styles to the stylesheet used by the guide (right now we include the the css on each HTML page).
-  Add logic to the doc build to automate the creation of widgets, or at least add some kind of validation to the PR builds.
-  (nice-to-have) Add a copy button for copying commands displayed in the tabs.

**Comment by gtback** [comment]
I don't want to put a lot of time into this if it will be changing with the new UI, but there are some things we can do now which would be re-usable.

> - Get design/developer resources to improve the javascript and css code. Right now, the tabbed panels display fine on desktop browsers, but aren't optimized for mobile (though that's true of our docs, too).

@Adrian-J do we have any existing patterns for this type of thing on the website? Would something like the **Tabbed View** on the [new design](https://marvelapp.com/e7g5fch/screen/69872623) work for this?

That should cover the CSS for now. For the JavaScript, I haven't looked at it and I'm not a JS expert, but I can probably find someone to help with that.

> - Add styles to the stylesheet used by the guide (right now we include the the css on each HTML page).

We can add them into into [our styles](https://github.com/elastic/docs/tree/master/resources/web/style), and I'll make sure they're part of the new design as well. 

> - Add logic to the doc build to automate the creation of widgets, or at least add some kind of validation to the PR builds.

This seems like a good use for an [Asciidoctor extension](https://asciidoctor.org/docs/user-manual/#extensions).

> - (nice-to-have) Add a copy button for copying commands displayed in the tabs.

Should be relatively easy to add via some front-end JavaScript (likely borrowing from the Console widget).

**Comment by gchaps** [comment]
++ to having to having this "widget".  We would also like to use it for the Install docs in the Kibana documentation.

**Comment by Adrian-J** [comment]
@gtback Yes, the Tabbed View component would be perfect here.

**Comment by gtback** [comment]
@bmorelli25 I think this one is good to close based on your recent work. Do you agree?

**Comment by bmorelli25** [comment]
Ah, I forgot about this issue! Linking in the PR where CSS/JS was added to the doc build:  https://github.com/elastic/docs/pull/2180. There's still some design work that we've been waiting many months for  (https://github.com/elastic/website-ux/issues/112), but the design team has recently stated they won't be helping with that. So yeah, we can probably close this.

**Comment by gtback** [comment]
(Since @bmorelli25 did all the work 😁 )
